storage/memory.go

```diff
@@ -1,62 +1,122 @@
 package storage
 
 import (
+	"encoding/gob"
 	"fmt"
-	"sync"
-
 	"github.com/TwinProduction/gatus/core"
+	"github.com/TwinProduction/gocache"
+	"io/ioutil"
+	"os"
+	"time"
 )
 
-// InMemoryStore implements an in-memory store
-type InMemoryStore struct {
-	serviceStatuses     map[string]*core.ServiceStatus
-	serviceResultsMutex sync.RWMutex
+// Store implements an in-memory store
+type Store struct {
+	serviceStatuses *gocache.Cache
+	interval        *time.Duration
+	filePath        string
+	shouldPersist   bool
 }
 
-// NewInMemoryStore returns an in-memory store. Note that the store acts as a singleton, so although new-ing
+// NewStore returns an in-memory store. Note that the store acts as a singleton, so although new-ing
 // up in-memory stores will give you a unique reference to a struct each time, all structs returned
 // by this function will act on the same in-memory store.
-func NewInMemoryStore() *InMemoryStore {
-	return &InMemoryStore{
-		serviceStatuses: make(map[string]*core.ServiceStatus),
+func NewStore() *Store {
+	return &Store{
+		serviceStatuses: gocache.NewCache(),
+	}
+}
+
+func init() {
+	gob.Register(&core.Result{})
+	gob.Register(&core.Uptime{})
+	gob.Register(&core.ServiceStatus{})
+}
+
+// WithPersistence configures the in-memory store to read and write to a given filepath at the specified interval.
+// If interval is nil then every cache write causes an immediate flush to disk.
+func (ims *Store) WithPersistence(filePath string, interval *time.Duration) *Store {
+	if _, err := os.Stat(filePath); os.IsNotExist(err) {
+		err := ioutil.WriteFile(filePath, []byte{}, 0644)
+		if err != nil {
+			panic(fmt.Sprintf("Could not write to file: %s", err))
+		}
+	} else {
+		numEvictions, err := ims.serviceStatuses.ReadFromFile(filePath)
+		if numEvictions != 0 {
+			panic(fmt.Sprintf("Unexpectedly dropped %d cache entries", numEvictions))
```

Comment:
Is it worth panicking here? This would prevent the application from ever recovering, because the next time the application starts, it'll read the file and probably panic again, looping until manual action is taken.

IMHO, since the usage of persistence on Gatus will likely be for convenience rather than a critical requirement.

I mean, technically, this shouldn't happen, but if it does, how bad can it be? 🤔

Reply:
No that is a fair point - it shouldn't happen as the cache is not itself size limited, but panicking does seem a bit excessive